### PR TITLE
Miscellaneous small fixes to superfluid logic

### PR DESCRIPTION
**Pull request by ValarDragon**
## Description

Minor core re-structure, solves launch blocking components of #830 and #864 

Documents, re-structures, and re-orders the logic at EpochEnd
______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: x/superfluid/keeper/gov.go [hunk 1]
```diff
@@ -1,34 +1,38 @@
...
-				sdk.NewAttribute(types.AttributeDenom, asset.Denom),
-				sdk.NewAttribute(types.AttributeSuperfluidAssetType, asset.AssetType.String()),
-			),
-		})
+		event := sdk.NewEvent(
+			types.TypeEvtSetSuperfluidAsset,
+			sdk.NewAttribute(types.AttributeDenom, asset.Denom),
+			sdk.NewAttribute(types.AttributeSuperfluidAssetType, asset.AssetType.String()),
+		)
+		ctx.EventManager().EmitEvent(event)
```
Comment (UnityChaos) [inline x/superfluid/keeper/gov.go:18]: I assume this is just a preferred style thing?

File: x/superfluid/keeper/hooks.go [hunk 2]
```diff
@@ -10,50 +11,73 @@ import (
 	"github.com/osmosis-labs/osmosis/x/superfluid/types"
 )
 
-func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochNumber int64) {
+func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, _ int64) {
```
Comment (UnityChaos) [inline x/superfluid/keeper/hooks.go:14]: Why not remove the arg entirely?

File: x/superfluid/keeper/hooks.go [hunk 3]
```diff
@@ -10,50 +11,73 @@ import (
...
+		bondDenom := k.sk.BondDenom(ctx)
+		osmoPoolAsset, err := pool.GetPoolAsset(bondDenom)
+		if err != nil {
+			// Pool has unexpectedly removed Osmo from its assets.
+			k.Logger(ctx).Error(err.Error())
+			k.BeginUnwindSuperfluidAsset(ctx, 0, asset)
+			return err
+		}
+
+		twap := osmoPoolAsset.Token.Amount.ToDec().Quo(pool.GetTotalShares().Amount.ToDec())
```
Comment (UnityChaos) [inline x/superfluid/keeper/hooks.go:69]: Maybe we should add a comment here indicating that we're faking TWAP with spot price.

**Review by UnityChaos — APPROVED** [review]
LGTM

File: x/superfluid/keeper/hooks.go (hunk 2 above)
Comment (ValarDragon) [reply x/superfluid/keeper/hooks.go:14]: Hooks must match interfaces.

File: x/superfluid/keeper/hooks.go (hunk 3 above)
Comment (ValarDragon) [reply x/superfluid/keeper/hooks.go:69]: Agreed

File: x/superfluid/keeper/gov.go (hunk 1 above)
Comment (ValarDragon) [reply x/superfluid/keeper/gov.go:18]: Yeah, and I want that event creation to be a different function later.

File: x/superfluid/keeper/hooks.go (hunk 3 above)
Comment (ValarDragon) [reply x/superfluid/keeper/hooks.go]: Resolved in latest commit

**Pull request merged**